coffea/processor/servicex/executor.py

```diff
@@ -0,0 +1,197 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from abc import ABC, abstractmethod
+from typing import Any, Callable, AsyncGenerator
+from urllib.parse import urlparse, unquote
+from urllib.request import url2pathname
+
+import aiostream
+import uproot
+from .accumulator import Accumulator
+from .analysis import Analysis
+from servicex import StreamInfoUrl
+
+
+class Executor(ABC):
+    @abstractmethod
+    def run_async_analysis(
+        self,
+        file_url: str,
+        tree_name: str,
+        accumulator: Accumulator,
+        process_func: Callable,
+    ):
+        raise NotImplementedError
+
+    def get_result_file_stream(self, datasource):
+        return datasource.stream_result_file_urls()
+
+    async def execute(self, analysis, datasource):
+        """
+        Launch an analysis against the given dataset on the implementation's task framework
+        :param analysis:
+            The analysis to run
+        :param datasource:
+            The datasource to run against
+        :return:
+            Stream of up to date histograms. Grows as each result is received
+        """
+        # Stream transformed file references from ServiceX
+        result_file_stream = self.get_result_file_stream(datasource)
+
+        # Launch a task against this file
+        func_results = self.launch_analysis_tasks_from_stream(
+            result_file_stream, analysis.accumulator, analysis.process
+        )
+
+        # Wait for all the data to show up
+        async def inline_wait(r):
+            "This could be inline, but python 3.6"
+            x = await r
+            return x
+
+        finished_events = aiostream.stream.map(func_results, inline_wait, ordered=False)
+        # Finally, accumulate!
+        # There is an accumulate pattern in the aiostream lib
+        output = analysis.accumulator.identity()
+        async with finished_events.stream() as streamer:
+            async for results in streamer:
```

Comment:
Since you want to yield the partial result each step, you'll have to lift https://github.com/CoffeaTeam/coffea/blob/master/coffea/processor/accumulator.py#L83-L97 and convert it to async (I think it would be enough to just change `next(gen)` to `await next(gen)` and then add a `yield` after line 94